src/engine/stack_mmap_XStream.c

```diff
@@ -0,0 +1,87 @@
+/**
+ * (C) Copyright 2016-2021 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+/**
+ * \file
+ *
+ * This file contains the engine/XStream specific part of the
+ * mmap()'ed ULT stack allocation feature.
+ */
+
+#ifdef ULT_MMAP_STACK
+#include <daos/common.h>
+#include "srv_internal.h"
+#include <errno.h>
+#include <string.h>
+
+/* XXX both thresholds may need to be dynamically determined based on the
+ * number of free stack pool (one per-XStream), and max_nb_mmap_stacks (see
+ * below).
+ */
+#define MAX_PERCENT_FREE_STACKS 20
+#define MAX_NUMBER_FREE_STACKS 2000
+
+/* callback to free stack upon ULT exit during stack_key deregister */
+void free_stack(void *arg)
+{
+	mmap_stack_desc_t *desc = (mmap_stack_desc_t *)arg;
+	struct dss_xstream *dx = dss_current_xstream();
+	struct stack_pool *sp;
+
+	/* XXX
+	 * We may need to reevaluate stack size since a growth may
+	 * have occurred during previous context life time, if initial
+	 * stack size has overflowed when there was no previous mapping
+	 * in address space to trigger Kernel's stack guard gap
+	 * (stack_guard_gap) ? This for both munmap() or queuing in
+	 * free pool cases.
+	 */
+
+	/* desc->sp should have been set to executing XStream's stacks
+	 * pool, but double-check just in case ULT has been migrated to an
+	 * other XStream in-between ...
+	 */
```

Comment:
I think it should be always put to the xstream where ULT is executed, DAOS doesn't have ULT migration use case.

Reply:
> I think it should be always put to the xstream where ULT is executed, DAOS doesn't have ULT migration use case.

Let's double-check anyway !!

Reply:
For double check, you'd put an assert (dx->dx_sp == desc->sp). Using 'desc->sp' when 'dx == NULL' looks incorrect to me.

Reply:
Ok, I think I see what is hurting you here now. The double-check I tried to indicate in the comment is not "dx == NULL" (only intended to handle some possible/early corner cases I triggered during my testing) but the fact the current XStream is again fetched and used.
So ok, I will add the (dx->dx_sp == desc->sp) real check you have indicated.